File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
res = [1 for _ in range(n)]
numbers = list(map(int, input().split()))
for i in range(n):
    for j in range(i):
        if numbers[i]>numbers[j] :
            res[i]=max(res[i], res[j]+1)
max_length= max(res)
print(max_length)
idx = res.index(max_length)
res_list = []
while idx>=0:
    if res[idx]==max_length:
        res_list.append(numbers[idx])
        max_length-=1
    idx -=1

for num in res_list[::-1]:
    print(num, end=' ')

File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func Find(arr []int, target int) int {
	for i, n := range arr {
		if target == n {
			return i
		}
	}
	return -1
}
func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n int
	fmt.Fscanf(reader, "%d\n", &n)
	var numbers = make([]int, n)
	var res = make([]int, n)
	var max int
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &numbers[i])
		res[i] = 1
	}
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			if numbers[i] > numbers[j] {
				if res[i] < res[j]+1 {
					res[i] = res[j] + 1
				}
			}
		}
	}
	for _, v := range res {
		if max < v {
			max = v
		}
	}
	fmt.Fprintf(writer, "%d\n", max)
	idx := Find(res, max)
	var resList = []int{}
	for idx >= 0 {
		if res[idx] == max {
			resList = append(resList, numbers[idx])
			max--
		}
		idx--
	}
	sort.Ints(resList)
	for _, v := range resList {
		fmt.Fprintf(writer, "%d ", v)
	}
}
